### DISCUSSION TITLE
Experimental feature: Update the size of the search queue

### DISCUSSION THREAD

**Discussion by irevoire** [meilisearch/product#729] [Experimental features]
<table>
    <thead>
        <tr>
            <th>Feature name</th>
  			<th>Enable when launching</th>
  			<th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>Customize the size of the search queue</td>
           <td>

```
./meilisearch --experimental-search-queue-size 10
```
  
</td>            
<td>Output logs as json</td>
        </tr>
    </tbody>
</table>

Meilisearch v1.8.0 (released on May 06, 2024) introduces a new experimental cli flag to control the number of search requests that can stay in a queue before processing.
The larger the queue is, the more RAM consumption it'll cause.
The default value is 1000. You might want to increase this number if you see very little RAM consumption on your server but receive HTTP 503 errors.
You might want to decrease this number if you notice that Meilisearch crashes during your search spike.

## Usage

```bash
./meilisearch --experimental-search-queue-size 10
```

**Comment by flexchar** [comment]
Just so I understand correctly that ONCE those search requests are processed the system resources such as RAM would be released back?

**Reply by irevoire** [reply]
Hey, yes don't worry 😁

**Comment by doutatsu** [comment]
How would I pass this when using a `Dockerfile`?

```
ARG MEILI_MASTER_KEY
ARG MEILI_VERSION

FROM getmeili/meilisearch:v${MEILI_VERSION}

EXPOSE 7700
```

**Reply by irevoire** [reply]
Hey @doutatsu,

You can use the following environment variable; `EXPERIMENTAL_SEARCH_QUEUE_SIZE`

**Reply by doutatsu** [reply]
@irevoire Do I need to also set the flag, e.g. `experimental-features', { 'vectorStore' => false }`? I've set this to 2000 but I don't see any changes (I assumed the RAM usage to increase)

**Reply by curquiza** [reply]
@doutatsu `vectorStore` experimental feature is not related with this one. You don't need to activate it except if you want to use hybrid search

**Reply by doutatsu** [reply]
@curquiza I am aware, I was asking if this feature also needs to be activated in a similar fashion. Cause simply setting the variable changes nothing

**Reply by curquiza** [reply]
Only activating the variable activates the feature. No need to use the API call

**Reply by doutatsu** [reply]
@curquiza I've followed what @irevoire suggested, but my RAM usage is the same - I would have assumed it to increase, as mentioned:

<img width="789" alt="image" src="https://github.com/meilisearch/product/assets/4270980/0194e696-a7c7-4132-93dc-c5057dcef0b3">
<img width="501" alt="image" src="https://github.com/meilisearch/product/assets/4270980/4cee5b0f-c59d-4fed-ac9b-c8f585999667">

**Reply by irevoire** [reply]
Hey @doutatsu, yeah the search queue is not supposed to consume much RAM.
First of all, with a default value of 1000 meilisearch, it will only use around 7.8KiB of RAM at startup.
In my tests, if 1000 search requests are processed concurrently, the size could reach 20MiB, but that only happens if the queue is full.

Increasing the value to 8000 should increase your RAM consumption to 62KiB and ~150MiB if the queue gets full.

That’s for the meilisearch side, but then, if you don't receive any search requests, your OS may lie about the allocation and wait until you receive something before actually consuming any RAM.

-----

Does that help?
Is it possible that you missed the allocation because you were not sending that many search requests?

**Reply by doutatsu** [reply]
@irevoire Ah I see, I was expecting going from 1000 to 2000 would double my existing average RAM usage - which is why I was looking for a huge spike. I am using this to try and see if it will help with timeouts, so I am going to keep observing for now in case I see any changes either in that or any RAM spikes. 

For reference, the average throughput is about 0.38 rps atm

**Reply by irevoire** [reply]
> I am using this to try and see if it will help with timeouts

Once the queue is full, meilisearch won't return a timeout but instead a 503, service unavailable. So it shouldn't be the case of your trouble.
Is there an issue with these timeouts? Maybe we can help you understand what's going on.
The first guess would be that they come from a reverse proxy in your infra 🤔

**Reply by doutatsu** [reply]
@irevoire It's extension of a long issue conversation in the rails integration - https://github.com/meilisearch/meilisearch-rails/issues/355. I was told to try this experimental feature.

There is no reverse proxy, I am simply using the rails client to make requests to the server

**Comment by Strijdhagen** [comment]
It appears that this setting defaults to 0 (possibly, only after enabling experimental features).
I just got the error in my application using javascript SDK: MeiliSearchApiError: 
Too many search requests running at the same time: 0. Retry after 10s.

Only after adding the experimental setting to my toml file did the error go away.

**Reply by irevoire** [reply]
Hey, I’m not sure to understand, can you share with me the way you enabled the feature?

**Reply by Nio-o** [reply]
I have same issue. After upgrading to 1.9.0 I have error "Too many search requests running at the same time: 0. Retry after 10s".

When I add argument `--experimental-search-queue-size 1000` to entrypoint script (or config.toml) error disappears. It seems like default value is not 1000

<details><summary>Config with error (actually default config from previous versions)</summary>

```
db_path = "/var/lib/meilisearch/data"
env = "production"
http_addr = "<REDACTED>"
master_key = "<REDACTED>"
no_analytics = true
http_payload_size_limit = "100 MB"
log_level = "INFO"
dump_dir = "/var/lib/meilisearch/dumps"
ignore_missing_dump = false
ignore_dump_if_db_exists = false
schedule_snapshot = false
snapshot_dir = "/var/lib/meilisearch/snapshots"
ignore_missing_snapshot = false
ignore_snapshot_if_db_exists = false
ssl_require_auth = false
ssl_resumption = false
ssl_tickets = false
experimental_enable_metrics = false
experimental_reduce_indexing_memory_usage = false
```

</details> 

Adding `experimental_search_queue_size = 1000` at the end helps to avoid described issue.

Test example:

```bash
npx autocannon \
  -H 'accept: application/json' \
  -H 'authorization: Bearer <TOKEN>' \
  -H 'content-type: application/json' \
  -b '{"queries":[<ANY QUERIES HERE>]}' \
  -c 10 \
  -d 10 \
  -m POST \
  --renderStatusCodes \
  https://example.com
```

Tested on 1.9.0 and 1.10.2

**Reply by irevoire** [reply]
Hey @Nio-o, thanks for the second report.

I investigated a bit more and [found the issue](https://github.com/meilisearch/meilisearch/issues/4991)! It’ll be fixed in the next rc

Thanks a lot @Nio-o and @Strijdhagen for finding and reporting the bug 🎉
